File: python/securekit/symmetric_encryptor.py
"""AEAD encryption using XChaCha20-Poly1305, via the `cryptography` package's
hazmat primitives -- no custom crypto.

Nonces are generated internally and prepended to the ciphertext; there is no
API to supply a caller-chosen nonce, eliminating nonce-reuse misuse.
"""

import base64
import os

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305

from .secure_random import secure_random_bytes

_KEY_LEN = 32
_NONCE_LEN = 12  # ChaCha20Poly1305 (IETF, 96-bit nonce); XChaCha20 not exposed
# by `cryptography`'s stable API, so we use the standard 96-bit-nonce variant
# with an internally generated random nonce per message (safe for the
# volume of messages any single key here is expected to encrypt).


class DecryptionError(Exception):
    """Raised for any decryption failure. Message is intentionally generic
    (never distinguishes "too short" vs "tag mismatch") to avoid leaking
    oracle information to an attacker."""

    def __init__(self) -> None:
        super().__init__("securekit: decryption failed")


class SymmetricEncryptor:
    def generate_key(self) -> bytes:
        return secure_random_bytes(_KEY_LEN)

    def encrypt(self, key: bytes, plaintext: bytes, aad: bytes | None = None) -> bytes:
        if len(key) != _KEY_LEN:
            raise ValueError("securekit: key must be 32 bytes")
        nonce = os.urandom(_NONCE_LEN)
        aead = ChaCha20Poly1305(key)
        ciphertext = aead.encrypt(nonce, plaintext, aad)
        return nonce + ciphertext

    def decrypt(self, key: bytes, blob: bytes, aad: bytes | None = None) -> bytes:
        if len(key) != _KEY_LEN or len(blob) < _NONCE_LEN:
            raise DecryptionError()
        nonce, ciphertext = blob[:_NONCE_LEN], blob[_NONCE_LEN:]
        aead = ChaCha20Poly1305(key)
        try:
            return aead.decrypt(nonce, ciphertext, aad)
        except InvalidTag as exc:
            raise DecryptionError() from exc

    def encrypt_to_string(self, key: bytes, plaintext: bytes, aad: bytes | None = None) -> str:
        return base64.b64encode(self.encrypt(key, plaintext, aad)).decode("ascii")

    def decrypt_from_string(self, key: bytes, encoded: str, aad: bytes | None = None) -> bytes:
        try:
            blob = base64.b64decode(encoded, validate=True)
        except Exception as exc:
            raise DecryptionError() from exc
        return self.decrypt(key, blob, aad)

File: php/src/SymmetricEncryptor.php
<?php

declare(strict_types=1);

namespace SecureKit;

/**
 * AEAD encryption using XChaCha20-Poly1305 via ext-sodium. Nonces are
 * generated internally and prepended to the ciphertext; there is no API to
 * supply a caller-chosen nonce, eliminating nonce-reuse misuse.
 */
final class SymmetricEncryptor
{
    private const KEY_LENGTH = SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_KEYBYTES; // 32
    private const NONCE_LENGTH = SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES; // 24

    public function generateKey(): string
    {
        return sodium_crypto_aead_xchacha20poly1305_ietf_keygen();
    }

    public function encrypt(string $key, string $plaintext, string $aad = ''): string
    {
        if (strlen($key) !== self::KEY_LENGTH) {
            throw new \InvalidArgumentException('securekit: key must be 32 bytes');
        }
        $nonce = random_bytes(self::NONCE_LENGTH);
        $ciphertext = sodium_crypto_aead_xchacha20poly1305_ietf_encrypt($plaintext, $aad, $nonce, $key);
        return $nonce . $ciphertext;
    }

    /**
     * @throws \RuntimeException with a generic message on any failure
     */
    public function decrypt(string $key, string $blob, string $aad = ''): string
    {
        if (strlen($key) !== self::KEY_LENGTH || strlen($blob) < self::NONCE_LENGTH) {
            throw new \RuntimeException('securekit: decryption failed');
        }
        $nonce = substr($blob, 0, self::NONCE_LENGTH);
        $ciphertext = substr($blob, self::NONCE_LENGTH);
        $plaintext = sodium_crypto_aead_xchacha20poly1305_ietf_decrypt($ciphertext, $aad, $nonce, $key);
        if ($plaintext === false) {
            throw new \RuntimeException('securekit: decryption failed');
        }
        return $plaintext;
    }

    public function encryptToString(string $key, string $plaintext, string $aad = ''): string
    {
        return base64_encode($this->encrypt($key, $plaintext, $aad));
    }

    public function decryptFromString(string $key, string $encoded, string $aad = ''): string
    {
        $blob = base64_decode($encoded, true);
        if ($blob === false) {
            throw new \RuntimeException('securekit: decryption failed');
        }
        return $this->decrypt($key, $blob, $aad);
    }
}
